shared/entity/type.go

```diff
@@ -0,0 +1,332 @@
+
+	// Always set project parameter if provided (operations and warnings may be project specific but it is not a requirement).
+	if projectName != "" {
+		u = u.WithQuery("project", projectName)
+	}
+
+	// Always set location if provided.
+	if location != "" {
+		u = u.WithQuery("target", location)
+	}
+
+	return Reference(u.String()), nil
+}
+
+// path returns the elements of the path that comprise the url of the Type. Placeholders are used where path arguments
+// are expected.
+func (t Type) path() ([]string, error) {
+	switch t {
+	case TypeContainer:
+		return []string{"containers", pathPlaceholder}, nil
+	case TypeImage:
+		return []string{"images", pathPlaceholder}, nil
+	case TypeProfile:
+		return []string{"profiles", pathPlaceholder}, nil
+	case TypeProject:
+		return []string{"projects", pathPlaceholder}, nil
+	case TypeCertificate:
+		return []string{"certificates", pathPlaceholder}, nil
+	case TypeInstance:
+		return []string{"instances", pathPlaceholder}, nil
+	case TypeInstanceBackup:
+		return []string{"instances", pathPlaceholder, "backups", pathPlaceholder}, nil
+	case TypeInstanceSnapshot:
+		return []string{"instances", pathPlaceholder, "snapshots", pathPlaceholder}, nil
+	case TypeNetwork:
+		return []string{"networks", pathPlaceholder}, nil
+	case TypeNetworkACL:
+		return []string{"network-acls", pathPlaceholder}, nil
+	case TypeNode:
+		return []string{"cluster", "members", pathPlaceholder}, nil
+	case TypeOperation:
+		return []string{"operations", pathPlaceholder}, nil
+	case TypeStoragePool:
+		return []string{"storage-pools", pathPlaceholder}, nil
+	case TypeStorageVolume:
+		return []string{"storage-pools", pathPlaceholder, "volumes", pathPlaceholder, pathPlaceholder}, nil
+	case TypeStorageVolumeBackup:
+		return []string{"storage-pools", pathPlaceholder, "volumes", pathPlaceholder, pathPlaceholder, "backups", pathPlaceholder}, nil
+	case TypeStorageVolumeSnapshot:
+		return []string{"storage-pools", pathPlaceholder, "volumes", pathPlaceholder, pathPlaceholder, "snapshots", pathPlaceholder}, nil
+	case TypeStorageBucket:
+		return []string{"storage-pools", pathPlaceholder, "buckets", pathPlaceholder}, nil
+	case TypeWarning:
+		return []string{"warnings", pathPlaceholder}, nil
+	case TypeClusterGroup:
+		return []string{"cluster", "groups", pathPlaceholder}, nil
+	default:
+		return nil, fmt.Errorf("Missing path definition for entity type %q", t)
+	}
+}
+
+// ParseURL parses a raw URL string and returns the Type, project, location, and path arguments (mux vars).
+//
+// Path arguments are returned in the order they are found in the URL. If there is no project query parameter and the
+// Type requires a project, then api.ProjectDefaultName is returned as the project name. The returned location is the
+// value of the "target" query parameter. All returned values are unescaped.
+//
+// Warning: The rawURL argument to this function must be URL encoded. This method is likely to fail otherwise.
```

Comment:
Is there an example where the URL passed here could be a valid URL and not be URL encoded that would cause problems parsing it?

Reply:
I'm wondering if this should take a *url.URL type and then make the caller parse/construct the struct, rather than embed the call to url.Parse here?

Reply:
I've removed the warning as I'm now passing in a `url.URL`.